## ANÁLISIS DE DATOS

Importar los paquetes necesarios

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re
import warnings
import pyexcel as p 
import random
warnings.filterwarnings("ignore")
from matplotlib.dates import DateFormatter

Cargar el conjunto de datos

In [2]:
inventario_inicial=pd.read_csv(r'C:\Codeproject\Recursos\Datos Originales\BegInvFINAL12312016.csv')
inventario_final=pd.read_csv(r'C:\Codeproject\Recursos\Datos Originales\EndInvFINAL12312016.csv')
compras=pd.read_csv(r'C:\Codeproject\Recursos\Datos Originales\PurchasesFINAL12312016.csv')
precio_compras=pd.read_csv(r'C:\Codeproject\Recursos\Datos Originales\2017PurchasePricesDec.csv')
factura_compras=pd.read_csv(r'C:\Codeproject\Recursos\Datos Originales\InvoicePurchases12312016.csv')
ventas=pd.read_csv(r'C:\Codeproject\Recursos\Datos Originales\SalesFINAL12312016.csv')

Para tener una idea general de los datos, se han impreso las primeros 5lineas de cada tabla

In [3]:
inventario_inicial.head()

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   
2  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
3  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
4  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   

    Size  onHand  Price   startDate  
0  750mL       8  12.99  2016-01-01  
1  750mL       7  10.99  2016-01-01  
2  750mL       6  36.99  2016-01-01  
3  750mL       3  38.99  2016-01-01  
4  750mL       6  34.99  2016-01-01

In [4]:
inventario_final.head()

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
2  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
3  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   
4  1_HARDERSFIELD_75      1  HARDERSFIELD     75    Three Olives Tomato Vodka   

    Size  onHand  Price     endDate  
0  750mL      11  12.99  2016-12-31  
1  750mL       7  36.99  2016-12-31  
2  750mL       7  38.99  2016-12-31  
3  750mL       4  34.99  2016-12-31  
4  750mL       7  14.99  2016-12-31

In [5]:
compras.head()

InventoryId  Store  Brand                   Description   Size  \
0    69_MOUNTMEND_8412     69   8412     Tequila Ocho Plata Fresno  750mL   
1     30_CULCHETH_5255     30   5255  TGI Fridays Ultimte Mudslide  1.75L   
2    34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
3  1_HARDERSFIELD_5255      1   5255  TGI Fridays Ultimte Mudslide  1.75L   
4    76_DONCASTER_2034     76   2034     Glendalough Double Barrel  750mL   

   VendorNumber                   VendorName  PONumber      PODate  \
0           105  ALTAMAR BRANDS LLC               8124  2015-12-21   
1          4466  AMERICAN VINTAGE BEVERAGE        8137  2015-12-22   
2          4466  AMERICAN VINTAGE BEVERAGE        8137  2015-12-22   
3          4466  AMERICAN VINTAGE BEVERAGE        8137  2015-12-22   
4           388  ATLANTIC IMPORTING COMPANY       8169  2015-12-24   

  ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0    2016-01-02  2016-01-04  2016-02-16          35.71         6   214.26   
1    2016-01-01  2016-01-07  2016-02-21           9.35         4    37.40   
2    2016-01-02  2016-01-07  2016-02-21           9.41         5    47.05   
3    2016-01-01  2016-01-07  2016-02-21           9.35         6    56.10   
4    2016-01-02  2016-01-09  2016-02-16          21.32         5   106.60   

   Classification  
0               1  
1               1  
2               1  
3               1  
4               1

In [6]:
precio_compras.head()

Brand                  Description  Price   Size Volume  Classification  \
0     58  Gekkeikan Black & Gold Sake  12.99  750mL    750               1   
1     62     Herradura Silver Tequila  36.99  750mL    750               1   
2     63   Herradura Reposado Tequila  38.99  750mL    750               1   
3     72         No. 3 London Dry Gin  34.99  750mL    750               1   
4     75    Three Olives Tomato Vodka  14.99  750mL    750               1   

   PurchasePrice  VendorNumber                   VendorName  
0           9.28          8320  SHAW ROSS INT L IMP LTD      
1          28.67          1128  BROWN-FORMAN CORP            
2          30.46          1128  BROWN-FORMAN CORP            
3          26.11          9165  ULTRA BEVERAGE COMPANY LLP   
4          10.94          7245  PROXIMO SPIRITS INC.

In [7]:
factura_compras.head()

VendorNumber                   VendorName InvoiceDate  PONumber  \
0           105  ALTAMAR BRANDS LLC           2016-01-04      8124   
1          4466  AMERICAN VINTAGE BEVERAGE    2016-01-07      8137   
2           388  ATLANTIC IMPORTING COMPANY   2016-01-09      8169   
3           480  BACARDI USA INC              2016-01-12      8106   
4           516  BANFI PRODUCTS CORP          2016-01-07      8170   

       PODate     PayDate  Quantity    Dollars  Freight Approval  
0  2015-12-21  2016-02-16         6     214.26     3.47      NaN  
1  2015-12-22  2016-02-21        15     140.55     8.57      NaN  
2  2015-12-24  2016-02-16         5     106.60     4.61      NaN  
3  2015-12-20  2016-02-05     10100  137483.78  2935.20      NaN  
4  2015-12-24  2016-02-12      1935   15527.25   429.20      NaN

In [8]:
ventas.head()

InventoryId  Store  Brand                 Description        Size  \
0  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
1  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
2  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
3  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
4  1_HARDERSFIELD_1005      1   1005     Maker's Mark Combo Pack  375mL 2 Pk   

   SalesQuantity  SalesDollars  SalesPrice SalesDate  Volume  Classification  \
0              1         16.49       16.49  1/1/2016     750               1   
1              2         32.98       16.49  1/2/2016     750               1   
2              1         16.49       16.49  1/3/2016     750               1   
3              1         14.49       14.49  1/8/2016     750               1   
4              2         69.98       34.99  1/9/2016     375               1   

   ExciseTax  VendorNo                   VendorName  
0       0.79     12546  JIM BEAM BRANDS COMPANY      
1       1.57     12546  JIM BEAM BRANDS COMPANY      
2       0.79     12546  JIM BEAM BRANDS COMPANY      
3       0.79     12546  JIM BEAM BRANDS COMPANY      
4       0.79     12546  JIM BEAM BRANDS COMPANY

### Análisis exploratorio de datos

A continuación, se ha realizado el análisis exploratorio de datos, obteniendo una tabla donde se tiene detalles clave sobre los conjuntos de datos, como numero de columnas y filas, tipos de datos, valores no informados... Esta visión global ayuda a identificar errores, contribuyendo a la posterior limpieza y preprocesamiento de los datos.

In [9]:
# Se ha creado una función donde devuelve una tabla con la info necesaria para obtener información de los datos.
def dataframe_info(df):
    report = pd.DataFrame(columns=['Columna', 'TipoDato', 'Filas', 'Valores Unicos', 'Missings', 'Missings (%)'])
    for column in df.columns:
        data_type = df[column].dtype
        rows = df[column].shape[0] 
        unique_count = df[column].nunique()
        missing_values = df[column].isnull().sum()
        missing_percentage = (missing_values / len(df)) * 100
        report.loc[len(report)] = [column, data_type, rows, unique_count,  missing_values, missing_percentage.round(4)]
    return report

In [10]:
dataframe_info(inventario_inicial)

Columna TipoDato   Filas  Valores Unicos  Missings  Missings (%)
0  InventoryId   object  206529          206529         0           0.0
1        Store    int64  206529              79         0           0.0
2         City   object  206529              67         0           0.0
3        Brand    int64  206529            8094         0           0.0
4  Description   object  206529            7291         0           0.0
5         Size   object  206529              41         0           0.0
6       onHand    int64  206529             474         0           0.0
7        Price  float64  206529             329         0           0.0
8    startDate   object  206529               1         0           0.0

In [11]:
dataframe_info(inventario_final)

Columna TipoDato   Filas  Valores Unicos  Missings  Missings (%)
0  InventoryId   object  224489          224489         0         0.000
1        Store    int64  224489              80         0         0.000
2         City   object  224489              67      1284         0.572
3        Brand    int64  224489            9653         0         0.000
4  Description   object  224489            8732         0         0.000
5         Size   object  224489              47         0         0.000
6       onHand    int64  224489             548         0         0.000
7        Price  float64  224489             354         0         0.000
8      endDate   object  224489               1         0         0.000

In [12]:
dataframe_info(compras)

Columna TipoDato    Filas  Valores Unicos  Missings  Missings (%)
0      InventoryId   object  2372474          245907         0        0.0000
1            Store    int64  2372474              80         0        0.0000
2            Brand    int64  2372474           10664         0        0.0000
3      Description   object  2372474            9652         0        0.0000
4             Size   object  2372474              51         3        0.0001
5     VendorNumber    int64  2372474             126         0        0.0000
6       VendorName   object  2372474             129         0        0.0000
7         PONumber    int64  2372474            5543         0        0.0000
8           PODate   object  2372474             319         0        0.0000
9    ReceivingDate   object  2372474             364         0        0.0000
10     InvoiceDate   object  2372474             373         0        0.0000
11         PayDate   object  2372474             382         0        0.0000
12   PurchasePrice  float64  2372474            2151         0        0.0000
13        Quantity    int64  2372474             686         0        0.0000
14         Dollars  float64  2372474           33567         0        0.0000
15  Classification    int64  2372474               2         0        0.0000

In [13]:
dataframe_info(precio_compras)

Columna TipoDato  Filas  Valores Unicos  Missings  Missings (%)
0           Brand    int64  12261           12261         0        0.0000
1     Description   object  12261           11114         1        0.0082
2           Price  float64  12261             380         0        0.0000
3            Size   object  12261              56         1        0.0082
4          Volume   object  12261              33         1        0.0082
5  Classification    int64  12261               2         0        0.0000
6   PurchasePrice  float64  12261            2314         0        0.0000
7    VendorNumber    int64  12261             131         0        0.0000
8      VendorName   object  12261             136         0        0.0000

In [14]:
dataframe_info(factura_compras)

Columna TipoDato  Filas  Valores Unicos  Missings  Missings (%)
0  VendorNumber    int64   5543             126         0        0.0000
1    VendorName   object   5543             129         0        0.0000
2   InvoiceDate   object   5543             373         0        0.0000
3      PONumber    int64   5543            5543         0        0.0000
4        PODate   object   5543             319         0        0.0000
5       PayDate   object   5543             382         0        0.0000
6      Quantity    int64   5543            2895         0        0.0000
7       Dollars  float64   5543            5226         0        0.0000
8       Freight  float64   5543            4052         0        0.0000
9      Approval   object   5543               1      5169       93.2528

In [15]:
dataframe_info(ventas)

Columna TipoDato    Filas  Valores Unicos  Missings  Missings (%)
0      InventoryId   object  1048575          170131         0           0.0
1            Store    int64  1048575              79         0           0.0
2            Brand    int64  1048575            7658         0           0.0
3      Description   object  1048575            6890         0           0.0
4             Size   object  1048575              40         0           0.0
5    SalesQuantity    int64  1048575             141         0           0.0
6     SalesDollars  float64  1048575            3426         0           0.0
7       SalesPrice  float64  1048575             273         0           0.0
8        SalesDate   object  1048575              60         0           0.0
9           Volume    int64  1048575              22         0           0.0
10  Classification    int64  1048575               2         0           0.0
11       ExciseTax  float64  1048575             468         0           0.0
12        VendorNo    int64  1048575             116         0           0.0
13      VendorName   object  1048575             117         0           0.0

La exploración de datos ayuda a identificar irregularidades que necesitan limpieza y preprocesamiento:

- Algunas tablas contienen missings que es necesario identificar y tratar adecuadamente.
- Las columnas de tamaño en todas las tablas muestran incoherencias debido a la variación de las unidades de medida. Algunos ejemplos son 750mL, Litro, 750mL + 2/,2 Pk, 50mL 4 Pk, 750mL 3 Pk. Es necesario convertirlas a una unidad de medida única y coherente.
- Las columnas Tamaño y Volumen llevan datos duplicados y una de ellas puede eliminar.
- El formato de fecha difiere entre tablas. Por ejemplo, la tabla de ventas utiliza d/m/aaaa, mientras que otras utilizan aaaa/mm/dd.
- Algunas columnas contienen espacios finales entre las entradas, como se ve en la columna VendorName de la tabla de compras (por ejemplo, ALTAMAR BRANDS LLC ,)

### Limpieza y preprocesamiento de datos
 

A. Tratamiento de Missings:

- Las tablas "inventario_inicial" y "ventas" no tienen valores faltantes, por lo que nos concentraremos en las otras cuatro tablas.

- (A.1) En la tabla "inventario_final" en la columna "city": De un total de 224.489 entradas, faltan 1.284 valores. Primero se ha observado si todos los nulos de ciudad pertenecen a la misma tienda, lo cual se ha visto que si.Por lo tanto, se le ha asignado el nombre de la ciudad que pertence a es tienda, obteniendo el datos de la tabla inventario inicial.

- (A.2) Missings en la tabla "compras" en la columna "size": Sólo faltan 3 valores de un total de 2.372.474 entradas, por lo que se ha decidiso eliminar esas filas. 

- (A.3) Missings en la tabla "Ventas" en la columna "Approval": Faltan informacion en 5.169 registros de un total de 5.543, lo que supone el 93% del total.  Dado el alto porcentaje de valores faltantes, se ha decidido eliminar la columna.

- (A.4) Missings en la tabla "precio_compras": Faltan valores individuales en las columnas "Description", "Size" y "Volume". En estos casos como estamos hablando de un solo registro en cada caso, se ha decidido elimar dichas filas.

In [16]:
# A.1
# Filtrar las filas donde 'City' es NaN
city_nan = inventario_final[inventario_final['City'].isna()]

# Obtener los números de tienda de aquellas ciudades que tienen NaN en 'City'
tienda_city_nan = city_nan['Store'].unique()
print(f"Números de tienda con 'City' NaN: {tienda_city_nan}")

# Verificar si alguna vez la tienda 46 tiene un valor diferente de NaN en 'City'
store_46_rows = inventario_final[inventario_final['Store'] == 46]
store_46_non_nan_city = store_46_rows['City'].notna().any()


if store_46_non_nan_city:
    print("La tienda 46 tiene al menos un valor no NaN en la columna 'City'.")
else:
    print("Todos los valores de 'City' para la tienda 46 son NaN.")

Números de tienda con 'City' NaN: [46]
Todos los valores de 'City' para la tienda 46 son NaN.


In [17]:
#Se ha comprobado cual es la tienda 46 en la tabla inventario incial, obteniendo el resultado TYWARDREATH
store_46_rows = inventario_inicial[inventario_inicial['Store'] == 46]
store_46_rows

InventoryId  Store         City  Brand  \
108343     46_TYWARDREATH_58     46  TYWARDREATH     58   
108344     46_TYWARDREATH_62     46  TYWARDREATH     62   
108345     46_TYWARDREATH_63     46  TYWARDREATH     63   
108346     46_TYWARDREATH_77     46  TYWARDREATH     77   
108347     46_TYWARDREATH_79     46  TYWARDREATH     79   
...                      ...    ...          ...    ...   
110218  46_TYWARDREATH_46763     46  TYWARDREATH  46763   
110219  46_TYWARDREATH_46764     46  TYWARDREATH  46764   
110220  46_TYWARDREATH_46826     46  TYWARDREATH  46826   
110221  46_TYWARDREATH_46829     46  TYWARDREATH  46829   
110222  46_TYWARDREATH_46830     46  TYWARDREATH  46830   

                         Description   Size  onHand  Price   startDate  
108343   Gekkeikan Black & Gold Sake  750mL       8  12.99  2016-01-01  
108344      Herradura Silver Tequila  750mL       2  36.99  2016-01-01  
108345    Herradura Reposado Tequila  750mL       4  38.99  2016-01-01  
108346   Three Olives Espresso Vodka  750mL       4  14.99  2016-01-01  
108347      Three Olives Loopy Vodka  750mL      12  14.99  2016-01-01  
...                              ...    ...     ...    ...         ...  
110218  Clayhouse Svgn Bl Paso Roble  750mL       5  11.99  2016-01-01  
110219  Clayhouse Adobe Red Paso Rbl  750mL       8  11.99  2016-01-01  
110220  LaBelle Winery Grnt St Apple  375mL       5  14.99  2016-01-01  
110221     Pacific Rim Yakima Gewurz  750mL       7  11.99  2016-01-01  
110222         Pacific Rim Sweet Rsl  750mL      13   9.99  2016-01-01  

[1880 rows x 9 columns]

In [18]:
#Tratamiento de los Missing

#A.1
inventario_final["City"] = inventario_final["City"].fillna("TYWARDREATH")

#A.2
compras = compras[compras['Size'].notna()]
#purchases = compras[compras['Size'].dropna()]

#A.3 
factura_compras = factura_compras.drop(['Approval'], axis=1)

#A.4
precio_compras.dropna(subset=["Description"], inplace=True)
precio_compras.dropna(subset=["Size"], inplace=True)
precio_compras.dropna(subset=["Volume"], inplace=True)

# Funcion para confirmar que se han eliminado los missings
def confirmar_missings(*datasets):
    for df_name, df in datasets:
        contador_missings = {}
        for idx, column_name in enumerate(df.columns):
            missing_count =df[column_name].isnull().sum()
            if missing_count > 0:
                contador_missings[column_name] = missing_count
        
        if len(contador_missings) == 0:
            print(f"Valores no informados en {df_name} = {len(contador_missings)}")
        else:
            for column_name, count in contador_missings.items():
                print(f"Columna: {column_name}, Valores no informados: {count}")
       
confirmar_missings(("Inventario inicial", inventario_inicial), ("Inventario final", inventario_final), ("Compras", compras), ("Factura compras", factura_compras), ("Precio compras", precio_compras),("Ventas", ventas))

Valores no informados en Inventario inicial = 0
Valores no informados en Inventario final = 0
Valores no informados en Compras = 0
Valores no informados en Factura compras = 0
Valores no informados en Precio compras = 0
Valores no informados en Ventas = 0


B. Tratamiento de duplicados

A continuación se ha comprobado si las bases de datos contienen valores duplicados.

In [19]:
# Comprobacion de duplicados
datasets = [inventario_inicial, inventario_final, compras, factura_compras, precio_compras, ventas]

for i, df in enumerate(datasets, start=1):
    duplicados = df.duplicated().any()
    if duplicados:
        print(f"El dataset {i} tiene duplicados.")
    else:
        print(f"El dataset {i} no tiene duplicados.")
    print("-" * 35)  # Separador entre datasets

El dataset 1 no tiene duplicados.
-----------------------------------
El dataset 2 no tiene duplicados.
-----------------------------------
El dataset 3 no tiene duplicados.
-----------------------------------
El dataset 4 no tiene duplicados.
-----------------------------------
El dataset 5 no tiene duplicados.
-----------------------------------
El dataset 6 no tiene duplicados.
-----------------------------------


C. Irregularidades en las entradas de datos.

La columna Size varía en todos las tablas, excepto en el conjunto de datos precio_compras. La información sobre el tamaño, que representa el volumen, se introduce en varios formatos, como litro, mililitro, onza, paquete (pk) y combinaciones de los mismos. Por lo tanto, estas medidas se tienen que convertir en un litro estándar unificado. Primero de todo se han comprobado los recuentos únicos. A los que tengan un solo recuento y no presenten un patrón, se les a asignado una tasa equivalente. A los que tengan varios recuentos y un patrón, se han convertido según los patrones.

In [20]:
# Se ha observado el patron, obteniendo un recuento unico de cada entrada de tamaño en todos os conjuntos de datos 
data_frames = [inventario_inicial, inventario_final, precio_compras, compras, ventas]
medidas = pd.concat([df["Size"] for df in data_frames])
medidas_unicas = medidas.unique()
size_counts = {}
for size in medidas_unicas:
    size_counts[size] = medidas[medidas == size].count()
medidas_unicas_count = pd.DataFrame({
    'Size': medidas_unicas,
    'Total Count': [size_counts[size] for size in medidas_unicas]})
print(medidas_unicas_count)

        Size  Total Count
0      750mL      2113362
1      Liter        73920
2      1.75L       809721
3       50mL       195549
4      375mL       139094
..       ...          ...
71  750mL  3            1
72    1500ml           11
73    9000mL            1
74     19.5L            4
75        9L            1

[76 rows x 2 columns]


A partir de la salida de recuentos únicos, se ha observado  que hay 51 formas diferentes de expresar medidas de volumen.
Se han utilizado expresiones regulares para recorrerlos y aplicar tipos de conversión. Se han agrupado en tres patrones:
1) Ejemplos: 750ml, 750mL, 750ml + 3/, 750 4p, 750 4pk, etc.
2) Patrones como Litro, L o l
3) Tamaños expresados en Oz.

Se ha creado una funcion para aplicar estas conversiones a todos los conjuntos de datos.

In [21]:
# Manejar las irregularidades de tamaño en todos los conjuntos de datos(datasets) 
def convert_to_liters(size):
    size = str(size).lower()
    # Convert 3/100ml and 5/2 -->  
    if "3/100ml" in size: return round(3 * 0.01, 2)  #3/100ml--> 3 packs of 100ml 
    elif '5/2 oz' in size: return round(10 * 0.0295735, 2)  #5/2 oz --> 5 packs of 1/2
    
    #pasar de galón a litro
    elif 'gal' in size:
        gal_value = float(re.search(r'\d+\.*\d*', size).group())
        return round(gal_value * 3.786, 2)   
    
    #patrones como 750 ml, 750 ml, 750 ml + 3/, 750 4p, 750 4pk, etc.
    elif 'ml' in size:
        if 'p' in size:
            ml_value, pack_value = re.search(r'(\d+\.*\d*)\s*m*l*\s*(\d*)\s*p*/*P*k*/*', size).groups()
            ml_value = float(ml_value)
            pack_value = float(pack_value) if pack_value else 1
            return round(ml_value * pack_value / 1000, 2)  # p or pk means pack--> multiply
        elif '+' in size:
            ml_value, pack_value = map(float, re.findall(r'\d+\.*\d*', size))
            return round((ml_value + (pack_value * 50)) / 1000, 2)  
        else:
            ml_value = float(re.search(r'\d+\.*\d*', size).group())
            return round(ml_value / 1000, 2)  
    
    # convertir Litro o L o l
    elif 'liter' in size or 'l' in size:
        if size == 'liter' or size == 'l': return 1.00
        else:
            liter_value = float(re.search(r'\d+\.*\d*', size).group())
            return round(liter_value, 2)  
    
    #onzas a litros    
    elif 'oz' in size:
        oz_value = float(re.search(r'\d+\.*\d*', size).group())
        return round(oz_value * 0.0295735, 2) 
    else:
        return None

datasets = [inventario_inicial, inventario_final, compras, precio_compras, ventas]

# Aplicar la función de conversión a cada dataset
for i, dataset in enumerate(datasets):
    dataset['Size'] = dataset['Size'].apply(convert_to_liters)
    datasets[i] = dataset


In [22]:
# Eliminar los espacios finales en las columnas especificadas
def remove_spaces(df, *columns):
    for col in columns:
        df[col] = df[col].str.strip()
    return df
inventario_inicial = remove_spaces(inventario_inicial, "City", "Description")
inventario_final = remove_spaces(inventario_final, "City", "Description")
compras= remove_spaces(compras, "Description", "VendorName")
factura_compras = remove_spaces(factura_compras, "VendorName")
precio_compras = remove_spaces(precio_compras, "Description", "VendorName")
ventas = remove_spaces(ventas, "Description", "VendorName")

# Renombrar 'vendorNo' a 'VendorNumber' en las ventas para mantener uniformidad entre las diferentes tablas.
ventas.rename(columns={'VendorNo': 'VendorNumber'}, inplace=True) 

Normalizar el formato de fecha de la columna especificada en un DataFrame: Las columnas de fecha de los conjuntos de datos Inventario inicial e Inventario final ya tienen formato de fecha, por lo que no necesitamos formatearlas. Formatearemos el resto.

In [23]:
def format_time(df, *columns):
    for col in columns:
        # Convertir las fechas a objetos datetime
        df[col] = pd.to_datetime(df[col], errors='coerce')
        # Formatear las fechas en el formato deseado
        df[col] = df[col].dt.strftime('%d/%m/%Y')
    return df

# Aplicar la función a los DataFrames correspondientes
inventario_inicial = format_time(inventario_inicial, "startDate")
inventario_final = format_time(inventario_final, "endDate")
compras = format_time(compras, "PODate", "ReceivingDate", "InvoiceDate", "PayDate")
factura_compras = format_time(factura_compras, "InvoiceDate", "PODate", "PayDate")
ventas = format_time(ventas, "SalesDate")


Tambien se ha visto que los nombres de las ciudades no son reales, por lo que se han sustituido por aquellos que si lo son. 
En el analisis se han utilizado unicamente los datos de inventario inicial, por lo que se ha realizado el cambio en dicha tabla.

In [24]:
city_counts_begin = inventario_inicial['City'].value_counts()
city_counts_df = pd.DataFrame({'City': city_counts_begin.index, 'Count': city_counts_begin.values})
city_counts_df.to_excel('city_counts_begin.xlsx', index=False)

In [25]:
city_counts_end = inventario_final['City'].value_counts()
city_counts_df = pd.DataFrame({'City': city_counts_end.index, 'Count': city_counts_end.values})
city_counts_df.to_excel('city_counts_end.xlsx', index=False)

Se han pasado ambos excels y se han juntado todas las ciudades (las que se encuentran en las dos tablas) en una columna llamada "Invented_city"

In [26]:
# Leer el archivo CSV con las ciudades inventadas y las ciudades reales
df_ciudades = pd.read_csv(r'C:\Codeproject\Recursos\Datos Originales\Ciudades.csv', sep=',', encoding='latin1')
df_ciudades.head()

Ciudad real      Estado Ciudad original
0  Nueva York  Nueva York       MOUNTMEND
1     Buffalo  Nueva York      EANVERNESS
2   Rochester  Nueva York       DONCASTER
3     Yonkers  Nueva York         HORNSEY
4      Albany  Nueva York       GOULCREST

In [27]:
# Reemplazar las ciudades inventadas con las ciudades reales
inventario_inicial['City'] = inventario_inicial['City'].replace(dict(zip(df_ciudades['Ciudad original'], df_ciudades['Ciudad real'])))
inventario_final['City'] = inventario_final['City'].replace(dict(zip(df_ciudades['Ciudad original'], df_ciudades['Ciudad real'])))

# Añadir una nueva columna con el "Estado"
inventario_inicial = inventario_inicial.merge(df_ciudades[['Ciudad real', 'Estado']], how='left', left_on='City', right_on='Ciudad real').drop(columns='Ciudad real')
inventario_final = inventario_final.merge(df_ciudades[['Ciudad real', 'Estado']], how='left', left_on='City', right_on='Ciudad real').drop(columns='Ciudad real')

Una vez realizado el cambio de ciudades, tambien se ha modificado la columna InventoryId, que procede de la columna ciudad.

In [28]:
# Modificar la columna 'Inventory_Id' en función de los cambios en 'city' y 'brand'
inventario_inicial['Store'] = inventario_inicial['Store'].astype(str)
inventario_inicial['Brand'] = inventario_inicial['Brand'].astype(str)
inventario_inicial['InventoryId'] = inventario_inicial['Store'] + '_' + inventario_inicial['City'] + '_' + inventario_inicial['Brand']

inventario_final['Store'] = inventario_final['Store'].astype(str)
inventario_final['Brand'] = inventario_final['Brand'].astype(str)
inventario_final['InventoryId'] = inventario_final['Store'] + '_' + inventario_final['City'] + '_' + inventario_final['Brand']

In [29]:
# Convertir 'Store' y 'Brand' a cadenas de caracteres
ventas['Store'] = ventas['Store'].astype(str)
ventas['Brand'] = ventas['Brand'].astype(str)

# Fusionar 'sales' y 'begin_inventory' en función de 'Store' y 'Brand'
merged_sales = pd.merge(ventas, inventario_inicial[['Store', 'Brand', 'InventoryId']], on=['Store', 'Brand'], how='left')

# Actualizar 'InventoryId' en 'sales' con los valores fusionados
ventas['InventoryId'] = merged_sales['InventoryId_y']


#Lo mismo para la tabla compras
compras['Store'] = compras['Store'].astype(str)
compras['Brand'] = compras['Brand'].astype(str)

# Fusionar 'sales' y 'begin_inventory' en función de 'Store' y 'Brand'
merged_sales = pd.merge(compras, inventario_inicial[['Store', 'Brand', 'InventoryId']], on=['Store', 'Brand'], how='left')

# Actualizar 'InventoryId' en 'sales' con los valores fusionados
compras['InventoryId'] = merged_sales['InventoryId_y']


Comprobar que se ha realizado el cambio correctamente

In [30]:
inventario_inicial.head()

InventoryId Store      City Brand                  Description  Size  \
0  1_Syracuse_58     1  Syracuse    58  Gekkeikan Black & Gold Sake  0.75   
1  1_Syracuse_60     1  Syracuse    60       Canadian Club 1858 VAP  0.75   
2  1_Syracuse_62     1  Syracuse    62     Herradura Silver Tequila  0.75   
3  1_Syracuse_63     1  Syracuse    63   Herradura Reposado Tequila  0.75   
4  1_Syracuse_72     1  Syracuse    72         No. 3 London Dry Gin  0.75   

   onHand  Price   startDate      Estado  
0       8  12.99  01/01/2016  Nueva York  
1       7  10.99  01/01/2016  Nueva York  
2       6  36.99  01/01/2016  Nueva York  
3       3  38.99  01/01/2016  Nueva York  
4       6  34.99  01/01/2016  Nueva York

In [31]:
inventario_final.head()

InventoryId Store      City Brand                  Description  Size  \
0  1_Syracuse_58     1  Syracuse    58  Gekkeikan Black & Gold Sake  0.75   
1  1_Syracuse_62     1  Syracuse    62     Herradura Silver Tequila  0.75   
2  1_Syracuse_63     1  Syracuse    63   Herradura Reposado Tequila  0.75   
3  1_Syracuse_72     1  Syracuse    72         No. 3 London Dry Gin  0.75   
4  1_Syracuse_75     1  Syracuse    75    Three Olives Tomato Vodka  0.75   

   onHand  Price     endDate      Estado  
0      11  12.99  31/12/2016  Nueva York  
1       7  36.99  31/12/2016  Nueva York  
2       7  38.99  31/12/2016  Nueva York  
3       4  34.99  31/12/2016  Nueva York  
4       7  14.99  31/12/2016  Nueva York

In [32]:
compras.head()

InventoryId Store Brand                   Description  Size  \
0  69_Nueva York_8412    69  8412     Tequila Ocho Plata Fresno  0.75   
1   30_Riverside_5255    30  5255  TGI Fridays Ultimte Mudslide  1.75   
2      34_Elmira_5215    34  5215  TGI Fridays Long Island Iced  1.75   
3     1_Syracuse_5255     1  5255  TGI Fridays Ultimte Mudslide  1.75   
4   76_Rochester_2034    76  2034     Glendalough Double Barrel  0.75   

   VendorNumber                  VendorName  PONumber      PODate  \
0           105          ALTAMAR BRANDS LLC      8124  21/12/2015   
1          4466   AMERICAN VINTAGE BEVERAGE      8137  22/12/2015   
2          4466   AMERICAN VINTAGE BEVERAGE      8137  22/12/2015   
3          4466   AMERICAN VINTAGE BEVERAGE      8137  22/12/2015   
4           388  ATLANTIC IMPORTING COMPANY      8169  24/12/2015   

  ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0    02/01/2016  04/01/2016  16/02/2016          35.71         6   214.26   
1    01/01/2016  07/01/2016  21/02/2016           9.35         4    37.40   
2    02/01/2016  07/01/2016  21/02/2016           9.41         5    47.05   
3    01/01/2016  07/01/2016  21/02/2016           9.35         6    56.10   
4    02/01/2016  09/01/2016  16/02/2016          21.32         5   106.60   

   Classification  
0               1  
1               1  
2               1  
3               1  
4               1

In [33]:
ventas.head()

InventoryId Store Brand                 Description  Size  \
0  1_Syracuse_1004     1  1004  Jim Beam w/2 Rocks Glasses  0.75   
1  1_Syracuse_1004     1  1004  Jim Beam w/2 Rocks Glasses  0.75   
2  1_Syracuse_1004     1  1004  Jim Beam w/2 Rocks Glasses  0.75   
3  1_Syracuse_1004     1  1004  Jim Beam w/2 Rocks Glasses  0.75   
4  1_Syracuse_1005     1  1005     Maker's Mark Combo Pack  0.75   

   SalesQuantity  SalesDollars  SalesPrice   SalesDate  Volume  \
0              1         16.49       16.49  01/01/2016     750   
1              2         32.98       16.49  02/01/2016     750   
2              1         16.49       16.49  03/01/2016     750   
3              1         14.49       14.49  08/01/2016     750   
4              2         69.98       34.99  09/01/2016     375   

   Classification  ExciseTax  VendorNumber               VendorName  
0               1       0.79         12546  JIM BEAM BRANDS COMPANY  
1               1       1.57         12546  JIM BEAM BRANDS COMPANY  
2               1       0.79         12546  JIM BEAM BRANDS COMPANY  
3               1       0.79         12546  JIM BEAM BRANDS COMPANY  
4               1       0.79         12546  JIM BEAM BRANDS COMPANY

## Crear tablas de dimensiones y tablas de hechos

D_PRODUCTO

In [34]:
# Seleccionar las columnas deseadas
D_producto = precio_compras[['Brand', 'Description','Size','Price']]

#Cambiar el nombre de las columnas
D_producto.columns=['IdProducto', 'Producto','Volumen','Precio unidad']

# Mirar si hay duplicados
D_producto.drop_duplicates(inplace=True)

# Definir los límites de los grupos y las etiquetas
bins = [0, 15, 25, 50, 250, 3000]
labels = ['Low cost', 'Economic', 'Intermediate', 'Premium', 'Luxury']

# Crear una nueva columna con los grupos personalizados
D_producto['Categoría'] = pd.cut(D_producto['Precio unidad'], bins=bins, labels=labels)

# Mostrar el nuevo DataFrame
D_producto

IdProducto                      Producto  Volumen  Precio unidad  \
0              58   Gekkeikan Black & Gold Sake     0.75          12.99   
1              62      Herradura Silver Tequila     0.75          36.99   
2              63    Herradura Reposado Tequila     0.75          38.99   
3              72          No. 3 London Dry Gin     0.75          34.99   
4              75     Three Olives Tomato Vodka     0.75          14.99   
...           ...                           ...      ...            ...   
12256       44917     Ferreira 10-Yr Tawny Port     0.75          24.99   
12257       44944     Sanford Santa Rita Pnt Nr     0.75          22.99   
12258       45016  Neal One Lane Bridg Cab Svgn     0.75          93.99   
12259       46011       Folonari Pnt Nr Venezie     1.50          12.99   
12260       46964     Calera Viognier Mt Harlan     0.75          35.99   

          Categoría  
0          Low cost  
1      Intermediate  
2      Intermediate  
3      Intermediate  
4          Low cost  
...             ...  
12256      Economic  
12257      Economic  
12258       Premium  
12259      Low cost  
12260  Intermediate  

[12260 rows x 5 columns]

D_CLIENTE

In [35]:
# Seleccionar las columnas deseadas
D_cliente = ventas[['VendorNumber', 'VendorName']]

#Eliminar duplicados
D_cliente.drop_duplicates(inplace=True)

#Cambiar el nombre de las columnas
D_cliente.columns=['IdCliente','Cliente']

# Ordenar el DataFrame por IdCliente de manera ascendente
D_cliente.sort_values(by='IdCliente', inplace=True)

# Resetear el índice para tener un índice limpio después de la ordenación
D_cliente.reset_index(drop=True, inplace=True)

# Mostrar el nuevo DataFrame
D_cliente

IdCliente                        Cliente
0            2  IRA GOLDMAN AND WILLIAMS, LLP
1          105             ALTAMAR BRANDS LLC
2          287           APPOLO VINEYARDS LLC
3          388     ATLANTIC IMPORTING COMPANY
4          480                BACARDI USA INC
..         ...                            ...
111      90057        WALPOLE MTN VIEW WINERY
112      90058              ZORVINO VINEYARDS
113      98450              Serralles Usa LLC
114     172662                SWEETWATER FARM
115     173357            TAMWORTH DISTILLING

[116 rows x 2 columns]

D_PROVEEDOR

In [36]:
# Selecciona las columnas deseadas
D_proveedor = compras[['VendorNumber', 'VendorName']]

#Eliminar duplicados
D_proveedor.drop_duplicates(subset=['VendorNumber'], inplace=True)

#Cambiar el nombre de las columnas
D_proveedor.columns=['IdProveedor','Proveedor']

# Ordenar el DataFrame por IdCliente de manera ascendente
D_proveedor.sort_values(by='IdProveedor', inplace=True)

# Resetear el índice para tener un índice limpio después de la ordenación
D_proveedor.reset_index(drop=True, inplace=True)

Se ha cogido el dato de transporte desde la tabla facturas

In [37]:
factura_compras.columns

Index(['VendorNumber', 'VendorName', 'InvoiceDate', 'PONumber', 'PODate',
       'PayDate', 'Quantity', 'Dollars', 'Freight'],
      dtype='object')

In [38]:
factura_compras.columns=['IdProveedor','Proveedor','Fecha factura', 'Numero pedido','Fecha pedido','Fecha pago','Cantidad','Coste','Transporte']

In [39]:
factura_compras['Transporte por unidad'] = factura_compras['Transporte']/factura_compras['Cantidad']

In [40]:
# Unir las tablas basadas en 'IdProveedor'
unir_df = pd.merge(D_proveedor, factura_compras[['IdProveedor', 'Transporte por unidad']],
                     on='IdProveedor', how='left')

# Eliminar duplicados en caso de que 'IdProveedor' pueda tener múltiples entradas en factura_compras
unir_df = unir_df.drop_duplicates(subset=['IdProveedor'])

# Mostrar el nuevo DataFrame D_PROVEEDOR actualizado
D_proveedor=unir_df

In [41]:
D_proveedor

IdProveedor                      Proveedor  Transporte por unidad
0               2  IRA GOLDMAN AND WILLIAMS, LLP               0.122000
4              54    AAPER ALCOHOL & CHEMICAL CO               0.480000
5              60        ADAMBA IMPORTS INTL INC               0.073235
40            105             ALTAMAR BRANDS LLC               0.578333
80            200      AMERICAN SPIRITS EXCHANGE               0.045833
...           ...                            ...                    ...
5402        98450              Serralles Usa LLC               0.661923
5456        99166          STARK BREWING COMPANY               0.104966
5468       172662                SWEETWATER FARM               0.103333
5497       173357            TAMWORTH DISTILLING               0.095000
5542       201359             FLAVOR ESSENCE INC               0.090000

[126 rows x 3 columns]

D_TIENDA

In [42]:
# Selecciona las columnas deseadas
D_tienda = inventario_inicial[['Store', 'City','Estado']]

#Eliminar duplicados
D_tienda.drop_duplicates(inplace=True)

#Cambiar el nombre de las columnas
D_tienda.columns=['IdTienda','Ciudad','Estado']

D_tienda

IdTienda        Ciudad      Estado
0             1      Syracuse  Nueva York
3255          2  Garden Grove  California
5944          3       Yonkers  Nueva York
6425          4       Buffalo  Nueva York
9296          5     Knoxville   Tennessee
...         ...           ...         ...
199422       75      Pasadena       Texas
201563       76     Rochester  Nueva York
206678       77    Fort Worth       Texas
209143       78       Lansing    Michigan
210368       79        Fresno  California

[79 rows x 3 columns]

H_VENTAS

In [43]:
ventas.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'SalesQuantity',
       'SalesDollars', 'SalesPrice', 'SalesDate', 'Volume', 'Classification',
       'ExciseTax', 'VendorNumber', 'VendorName'],
      dtype='object')

In [72]:
H_ventas = ventas[['InventoryId','Store', 'Brand', 'VendorNumber', 'SalesDate','SalesQuantity','SalesPrice', 'SalesDollars','ExciseTax']]
H_ventas.columns=['IdInventario','IdTienda','IdProducto','IdCliente','Fecha venta','Unidades vendidas','Precio venta unidad','Precio sin impuesto','Impuesto']
H_ventas['Ventas']= H_ventas['Unidades vendidas'] * (H_ventas['Precio venta unidad'] + H_ventas['Impuesto'])
H_ventas['Fecha venta'] = pd.to_datetime(H_ventas['Fecha venta'], format='%d/%m/%Y')
H_ventas

IdInventario IdTienda IdProducto  IdCliente Fecha venta  \
0            1_Syracuse_1004        1       1004      12546  2016-01-01   
1            1_Syracuse_1004        1       1004      12546  2016-01-02   
2            1_Syracuse_1004        1       1004      12546  2016-01-03   
3            1_Syracuse_1004        1       1004      12546  2016-01-08   
4            1_Syracuse_1005        1       1005      12546  2016-01-09   
...                      ...      ...        ...        ...         ...   
1048570  15_Los Angeles_3887       19      39384       2000  2016-02-12   
1048571  15_Los Angeles_3887       19      39384       2000  2016-02-13   
1048572  15_Los Angeles_3887       19      39384       2000  2016-02-19   
1048573  15_Los Angeles_3887       19      39384       2000  2016-02-22   
1048574  15_Los Angeles_3887       19      39384       2000  2016-02-24   

         Unidades vendidas  Precio venta unidad  Precio sin impuesto  \
0                        1                16.49                16.49   
1                        2                16.49                32.98   
2                        1                16.49                16.49   
3                        1                14.49                14.49   
4                        2                34.99                69.98   
...                    ...                  ...                  ...   
1048570                  4                12.95                51.80   
1048571                  8                12.95               103.60   
1048572                  1                12.95                12.95   
1048573                  3                12.95                38.85   
1048574                  2                12.95                25.90   

         Impuesto  Ventas  
0            0.79   17.28  
1            1.57   36.12  
2            0.79   17.28  
3            0.79   15.28  
4            0.79   71.56  
...           ...     ...  
1048570      0.45   53.60  
1048571      0.90  110.80  
1048572      0.11   13.06  
1048573      0.34   39.87  
1048574      0.22   26.34  

[1048575 rows x 10 columns]

In [45]:
compras.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'VendorNumber',
       'VendorName', 'PONumber', 'PODate', 'ReceivingDate', 'InvoiceDate',
       'PayDate', 'PurchasePrice', 'Quantity', 'Dollars', 'Classification'],
      dtype='object')

In [46]:
H_compras = compras[['InventoryId','Store','Brand','VendorNumber','PODate','ReceivingDate','InvoiceDate','PayDate','PurchasePrice','Quantity','Dollars']]
H_compras.columns=['IdInventario','IdTienda','IdProducto','IdProveedor','Fecha orden de compra','Fecha llegada','Fecha factura','Fecha de pago','Precio compra unidad','Unidades compradas','Coste compras']

# Convertir las columnas 'Fecha llegada' y 'Fecha orden de compra' a formato datetime
H_compras['Fecha llegada'] = pd.to_datetime(H_compras['Fecha llegada'], format='%d/%m/%Y')
H_compras['Fecha orden de compra'] = pd.to_datetime(H_compras['Fecha orden de compra'], format='%d/%m/%Y') 
H_compras['Fecha factura'] = pd.to_datetime(H_compras['Fecha factura'], format='%d/%m/%Y') 
H_compras['Fecha de pago'] = pd.to_datetime(H_compras['Fecha de pago'], format='%d/%m/%Y') 

# Calcular el 'Plazo de entrega' restando las dos columnas de fechas
H_compras['Plazo de entrega'] = H_compras['Fecha llegada'] - H_compras['Fecha orden de compra']
H_compras

IdInventario IdTienda IdProducto  IdProveedor  \
0        69_Nueva York_8412       69       8412          105   
1         30_Riverside_5255       30       5255         4466   
2            34_Elmira_5215       34       5215         4466   
3           1_Syracuse_5255        1       5255         4466   
4         76_Rochester_2034       76       2034          388   
...                     ...      ...        ...          ...   
2372469    20_Mesquite_3348       49      22298        90058   
2372470   73_Rochester_5246        1      19556        90058   
2372471  41_Sacramento_7978       66      22297        90058   
2372472       9_Austin_2258       69      19557        90058   
2372473   55_Oceanside_4223       55      22298        90058   

        Fecha orden de compra Fecha llegada Fecha factura Fecha de pago  \
0                  2015-12-21    2016-01-02    2016-01-04    2016-02-16   
1                  2015-12-22    2016-01-01    2016-01-07    2016-02-21   
2                  2015-12-22    2016-01-02    2016-01-07    2016-02-21   
3                  2015-12-22    2016-01-01    2016-01-07    2016-02-21   
4                  2015-12-24    2016-01-02    2016-01-09    2016-02-16   
...                       ...           ...           ...           ...   
2372469            2016-12-19    2016-12-28    2017-01-09    2017-02-06   
2372470            2016-12-19    2016-12-27    2017-01-09    2017-02-06   
2372471            2016-12-19    2016-12-26    2017-01-09    2017-02-06   
2372472            2016-12-19    2016-12-26    2017-01-09    2017-02-06   
2372473            2016-12-19    2016-12-28    2017-01-09    2017-02-06   

         Precio compra unidad  Unidades compradas  Coste compras  \
0                       35.71                   6         214.26   
1                        9.35                   4          37.40   
2                        9.41                   5          47.05   
3                        9.35                   6          56.10   
4                       21.32                   5         106.60   
...                       ...                 ...            ...   
2372469                  8.06                  12          96.72   
2372470                  9.39                  12         112.68   
2372471                  6.75                  12          81.00   
2372472                  9.39                  12         112.68   
2372473                  8.06                  12          96.72   

        Plazo de entrega  
0                12 days  
1                10 days  
2                11 days  
3                10 days  
4                 9 days  
...                  ...  
2372469           9 days  
2372470           8 days  
2372471           7 days  
2372472           7 days  
2372473           9 days  

[2372471 rows x 12 columns]

Como la información de ventas se tiene unicamente para los meses de enero y febrero, se ha decidido filtrar la información de compras también a dichos meses

In [47]:
# Filtrar las filas por el rango de fechas deseado
fecha_inicio = '01/01/2016'
fecha_fin = '29/02/2016'
H_compras_filtrada = H_compras[(H_compras['Fecha llegada'] >= fecha_inicio) & (H_compras['Fecha llegada'] <= fecha_fin)]

# Mostrar el DataFrame filtrado
H_compras_filtrada

IdInventario IdTienda IdProducto  IdProveedor  \
0        69_Nueva York_8412       69       8412          105   
1         30_Riverside_5255       30       5255         4466   
2            34_Elmira_5215       34       5215         4466   
3           1_Syracuse_5255        1       5255         4466   
4         76_Rochester_2034       76       2034          388   
...                     ...      ...        ...          ...   
351643  69_Nueva York_20383       20      31465         9815   
351650      66_Buffalo_4702       64      13723         9815   
351653      67_Buffalo_2280       74       6523         9815   
351673       33_Yonkers_782       33      14706         9815   
351691       34_Elmira_4732       74      25126        90058   

       Fecha orden de compra Fecha llegada Fecha factura Fecha de pago  \
0                 2015-12-21    2016-01-02    2016-01-04    2016-02-16   
1                 2015-12-22    2016-01-01    2016-01-07    2016-02-21   
2                 2015-12-22    2016-01-02    2016-01-07    2016-02-21   
3                 2015-12-22    2016-01-01    2016-01-07    2016-02-21   
4                 2015-12-24    2016-01-02    2016-01-09    2016-02-16   
...                      ...           ...           ...           ...   
351643            2016-02-22    2016-02-29    2016-03-13    2016-04-08   
351650            2016-02-22    2016-02-29    2016-03-13    2016-04-08   
351653            2016-02-22    2016-02-29    2016-03-13    2016-04-08   
351673            2016-02-22    2016-02-29    2016-03-13    2016-04-08   
351691            2016-02-26    2016-02-29    2016-03-14    2016-04-24   

        Precio compra unidad  Unidades compradas  Coste compras  \
0                      35.71                   6         214.26   
1                       9.35                   4          37.40   
2                       9.41                   5          47.05   
3                       9.35                   6          56.10   
4                      21.32                   5         106.60   
...                      ...                 ...            ...   
351643                 10.38                   3          31.14   
351650                 11.56                   3          34.68   
351653                 10.06                   4          40.24   
351673                  4.60                  12          55.20   
351691                  6.62                  12          79.44   

       Plazo de entrega  
0               12 days  
1               10 days  
2               11 days  
3               10 days  
4                9 days  
...                 ...  
351643           7 days  
351650           7 days  
351653           7 days  
351673           7 days  
351691           3 days  

[319970 rows x 12 columns]

H_inventario

Se han cogifo los datos del inventario inicial ya que para el analisis se necesitan estos

In [48]:
H_inventario = inventario_inicial[['Store','Brand','onHand', 'startDate']]
H_inventario.columns=['IdTienda','IdProducto','Stock','Fecha']
H_inventario['Fecha'] = pd.to_datetime(H_inventario['Fecha'], format='%d/%m/%Y')
H_inventario

IdTienda IdProducto  Stock      Fecha
0             1         58      8 2016-01-01
1             1         60      7 2016-01-01
2             1         62      6 2016-01-01
3             1         63      3 2016-01-01
4             1         72      6 2016-01-01
...         ...        ...    ...        ...
213529       79      46985     13 2016-01-01
213530       79      47014     13 2016-01-01
213531       79      47090     19 2016-01-01
213532       79      90011     12 2016-01-01
213533       79      90089     24 2016-01-01

[213534 rows x 4 columns]

H_MARGEN

In [49]:
H_ventas.columns

Index(['IdInventario', 'IdTienda', 'IdProducto', 'IdCliente', 'Fecha venta',
       'Unidades vendidas', 'Precio venta unidad', 'Ventas', 'Impuesto',
       'Precio total'],
      dtype='object')

In [50]:
H_compras_filtrada.columns

Index(['IdInventario', 'IdTienda', 'IdProducto', 'IdProveedor',
       'Fecha orden de compra', 'Fecha llegada', 'Fecha factura',
       'Fecha de pago', 'Precio compra unidad', 'Unidades compradas',
       'Coste compras', 'Plazo de entrega'],
      dtype='object')

In [51]:
# Filtrar H_ventas y H_compras_filtrada para reducir el tamaño
H_ventas_filtered = H_ventas[H_ventas['IdProducto'].isin(H_compras_filtrada['IdProducto'])]
H_compras_filtrada_filtered = H_compras_filtrada[H_compras_filtrada['IdProducto'].isin(H_ventas['IdProducto'])]


In [52]:
H_ventas_filtered.drop(columns=['IdInventario','Ventas','Precio total'], inplace=True)

In [53]:
H_compras_filtrada_filtered.drop(columns=['IdInventario','Fecha orden de compra','Fecha factura','Fecha de pago','Coste compras'], inplace=True)

In [54]:
H_ventas_filtered.columns

Index(['IdTienda', 'IdProducto', 'IdCliente', 'Fecha venta',
       'Unidades vendidas', 'Precio venta unidad', 'Impuesto'],
      dtype='object')

In [55]:
H_compras_filtrada_filtered.columns

Index(['IdTienda', 'IdProducto', 'IdProveedor', 'Fecha llegada',
       'Precio compra unidad', 'Unidades compradas', 'Plazo de entrega'],
      dtype='object')

Al encontrarme con este error "MemoryError: Unable to allocate 1.50 GiB for an array with shape (200729188,) and data type int64" he tenido que solucionar de la siguiente manera:

In [56]:
# Definir el tamaño del chunk
chunksize = 10000  # ajusta este valor según tus necesidades

# Generar chunks para ambos DataFrames
H_ventas_chunks = [H_ventas_filtered[i:i+chunksize] for i in range(0, len(H_ventas_filtered), chunksize)]
H_compras_chunks = [H_compras_filtrada_filtered[i:i+chunksize] for i in range(0, len(H_compras_filtrada_filtered), chunksize)]

# Procesar los datos en lotes y luego unirlos
merged_chunks = []
for chunk_ventas, chunk_compras in zip(H_ventas_chunks, H_compras_chunks):
    merged_chunk = pd.merge(chunk_ventas, chunk_compras, on='IdProducto', how='inner')
    merged_chunks.append(merged_chunk)

# Concatenar los resultados de los lotes
merged_data = pd.concat(merged_chunks)

In [57]:
merged_data

IdTienda_x IdProducto  IdCliente Fecha venta  Unidades vendidas  \
0              1       1005      12546  2016-01-09                  2   
1              1       1005      12546  2016-01-09                  2   
2              1       1005      12546  2016-01-09                  2   
3              1       1005      12546  2016-01-15                  1   
4              1       1005      12546  2016-01-15                  1   
...          ...        ...        ...         ...                ...   
53747         38       1760       3664  2016-01-23                  1   
53748         38       1760       3664  2016-01-23                  1   
53749         38       1760       3664  2016-01-23                  1   
53750         38       1760       3664  2016-01-23                  1   
53751         38       1760       3664  2016-01-23                  1   

       Precio venta unidad  Impuesto IdTienda_y  IdProveedor Fecha llegada  \
0                    34.99      0.79          1        12546    2016-01-01   
1                    34.99      0.79         66        12546    2016-01-03   
2                    34.99      0.79         38        12546    2016-01-03   
3                    34.99      0.39          1        12546    2016-01-01   
4                    34.99      0.39         66        12546    2016-01-03   
...                    ...       ...        ...          ...           ...   
53747                25.99      1.84         50         3664    2016-02-22   
53748                25.99      1.84         39         3664    2016-02-25   
53749                25.99      1.84         64         3664    2016-02-25   
53750                25.99      1.84         34         3664    2016-02-26   
53751                25.99      1.84         34         3664    2016-02-23   

       Precio compra unidad  Unidades compradas Plazo de entrega  
0                     27.34                   6          10 days  
1                     27.34                  21          12 days  
2                     27.34                  14          12 days  
3                     27.34                   6          10 days  
4                     27.34                  21          12 days  
...                     ...                 ...              ...  
53747                 19.69                   6           4 days  
53748                 19.69                   6           7 days  
53749                 19.69                   5           7 days  
53750                 19.69                   2           8 days  
53751                 19.69                   5           5 days  

[2053327 rows x 13 columns]

In [58]:
merged_data['Margen bruto unidad']=merged_data['Precio venta unidad']-merged_data['Precio compra unidad']
merged_data['Ventas']= merged_data['Unidades vendidas'] * (merged_data['Precio venta unidad'] + merged_data['Impuesto'])
merged_data['Costes']= merged_data['Unidades compradas'] * merged_data['Precio compra unidad']

In [59]:
merged_data

IdTienda_x IdProducto  IdCliente Fecha venta  Unidades vendidas  \
0              1       1005      12546  2016-01-09                  2   
1              1       1005      12546  2016-01-09                  2   
2              1       1005      12546  2016-01-09                  2   
3              1       1005      12546  2016-01-15                  1   
4              1       1005      12546  2016-01-15                  1   
...          ...        ...        ...         ...                ...   
53747         38       1760       3664  2016-01-23                  1   
53748         38       1760       3664  2016-01-23                  1   
53749         38       1760       3664  2016-01-23                  1   
53750         38       1760       3664  2016-01-23                  1   
53751         38       1760       3664  2016-01-23                  1   

       Precio venta unidad  Impuesto IdTienda_y  IdProveedor Fecha llegada  \
0                    34.99      0.79          1        12546    2016-01-01   
1                    34.99      0.79         66        12546    2016-01-03   
2                    34.99      0.79         38        12546    2016-01-03   
3                    34.99      0.39          1        12546    2016-01-01   
4                    34.99      0.39         66        12546    2016-01-03   
...                    ...       ...        ...          ...           ...   
53747                25.99      1.84         50         3664    2016-02-22   
53748                25.99      1.84         39         3664    2016-02-25   
53749                25.99      1.84         64         3664    2016-02-25   
53750                25.99      1.84         34         3664    2016-02-26   
53751                25.99      1.84         34         3664    2016-02-23   

       Precio compra unidad  Unidades compradas Plazo de entrega  \
0                     27.34                   6          10 days   
1                     27.34                  21          12 days   
2                     27.34                  14          12 days   
3                     27.34                   6          10 days   
4                     27.34                  21          12 days   
...                     ...                 ...              ...   
53747                 19.69                   6           4 days   
53748                 19.69                   6           7 days   
53749                 19.69                   5           7 days   
53750                 19.69                   2           8 days   
53751                 19.69                   5           5 days   

       Margen bruto unidad  Ventas  Costes  
0                     7.65   71.56  164.04  
1                     7.65   71.56  574.14  
2                     7.65   71.56  382.76  
3                     7.65   35.38  164.04  
4                     7.65   35.38  574.14  
...                    ...     ...     ...  
53747                 6.30   27.83  118.14  
53748                 6.30   27.83  118.14  
53749                 6.30   27.83   98.45  
53750                 6.30   27.83   39.38  
53751                 6.30   27.83   98.45  

[2053327 rows x 16 columns]

In [60]:
H_transacciones = merged_data

In [61]:
H_stock=merged_data

In [62]:
H_stock.drop_duplicates(subset=['IdProducto'])

IdTienda_x IdProducto  IdCliente Fecha venta  Unidades vendidas  \
0              1       1005      12546  2016-01-09                  2   
12             1      10058       2000  2016-01-05                  4   
75             1      10062      10754  2016-01-01                  1   
88             1       1013      17035  2016-01-06                  1   
90             1       1019       3664  2016-01-09                  1   
...          ...        ...        ...         ...                ...   
53548         38      17188      10754  2016-01-26                 12   
53654         38      17351       4425  2016-01-09                  1   
53655         38      17433      10754  2016-01-28                  1   
53677         38      17487       3252  2016-01-09                  1   
53690         38      17596       1601  2016-01-04                  2   

       Precio venta unidad  Impuesto IdTienda_y  IdProveedor Fecha llegada  \
0                    34.99      0.79          1        12546    2016-01-01   
12                   14.99      0.45         54         2000    2016-01-02   
75                    8.99      0.11         67        10754    2016-01-02   
88                   18.99      0.79         66        17035    2016-01-03   
90                   26.99      0.79         38         3664    2016-01-02   
...                    ...       ...        ...          ...           ...   
53548                16.99      1.35         76        10754    2016-02-29   
53654                16.99      0.11         34         4425    2016-02-29   
53655                13.99      0.11         34        10754    2016-02-29   
53677                 6.99      0.11         76         3252    2016-02-29   
53690                13.99      0.22         67         1601    2016-02-22   

       Precio compra unidad  Unidades compradas Plazo de entrega  \
0                     27.34                   6          10 days   
12                     9.26                  12           9 days   
75                     6.03                  10          13 days   
88                    10.29                   2          10 days   
90                    20.60                   1          12 days   
...                     ...                 ...              ...   
53548                 11.40                  12           7 days   
53654                 11.18                  12           3 days   
53655                  7.79                  12           7 days   
53677                  4.60                  12           7 days   
53690                  8.10                   3           3 days   

       Margen bruto unidad  Ventas  Costes  
0                     7.65   71.56  164.04  
12                    5.73   61.76  111.12  
75                    2.96    9.10   60.30  
88                    8.70   19.78   20.58  
90                    6.39   27.78   20.60  
...                    ...     ...     ...  
53548                 5.59  220.08  136.80  
53654                 5.81   17.10  134.16  
53655                 6.20   14.10   93.48  
53677                 2.39    7.10   55.20  
53690                 5.89   28.42   24.30  

[3605 rows x 16 columns]

In [63]:
H_stock=H_stock[['IdProducto','Unidades vendidas','Unidades compradas']]

In [64]:
# Definir el tamaño del chunk
chunksize = 10000  # ajusta este valor según tus necesidades

# Generar chunks para ambos DataFrames
H_stock_chunks = [H_stock[i:i+chunksize] for i in range(0, len(H_stock), chunksize)]
H_inventario_chunks = [H_inventario[['IdProducto', 'Stock']][i:i+chunksize] for i in range(0, len(H_inventario), chunksize)]

# Procesar los datos en lotes y luego unirlos
merged_chunks = []
for chunk_stock, chunk_inventario in zip(H_stock_chunks, H_inventario_chunks):
    merged_chunk = pd.merge(chunk_stock, chunk_inventario, on='IdProducto', how='left')
    merged_chunks.append(merged_chunk)

# Concatenar los resultados de los lotes
H_stock = pd.concat(merged_chunks)

In [65]:
H_stock = H_stock.groupby('IdProducto').sum()

In [71]:
D_cliente.to_excel("D_clientes.xlsx", index=False, engine='xlsxwriter')
D_proveedor.to_excel("D_proveedores.xlsx", index=False, engine='xlsxwriter')
D_producto.to_excel("D_productos.xlsx", index=False, engine='xlsxwriter')
D_tienda.to_excel("D_tiendas.xlsx", index=False, engine='xlsxwriter')

In [73]:
H_ventas.to_excel("H_ventas.xlsx", index=False, engine='xlsxwriter')
H_inventario.to_excel("H_inventario.xlsx", index=False, engine='xlsxwriter')
H_stock.to_excel("H_stock.xlsx", index=False, engine='xlsxwriter')

Como el archivo de compras es demasiado grande se ha intentado leer con un interfaz mas sencilla.

In [69]:
#Convertir el df a un diccionario de Python
data_dict = H_compras_filtrada.to_dict(orient='records')

#Guardar el diccionario en un archivo excel
p.save_as(records=data_dict, dest_file_name="H_compras_filtrada.xlsx")

In [70]:
#Convertir el df a un diccionario de Python
data_dict_2 = H_transacciones.to_dict(orient='records')

#Guardar el diccionario en un archivo excel
p.save_as(records=data_dict_2, dest_file_name="H_transacciones.xlsx")